feat: musubi_tuner as package

xhiroga opened the pull request:
Thank you for taking the time to review this despite your busy schedule...!

Regarding the packaging of musubi-tuner, I tried it out and found it to be more straightforward than expected. Therefore, this is a PR for sharing purposes!

It's perfectly fine if you close this with no plans to merge it for now! I'd be happy if you could take a look at it casually when you have some free time.

## About Packaging

Packaging actually just requires moving the code to a prescribed layout※1.

- Creating `setup.py` → Not necessary. With the adoption of `uv`, all necessary information is already included in `pyproject.toml`
- Registration to PyPI → Not mandatory. You can `pip install` directly from GitHub. Branch specification is also possible, by the way

The contents of this PR can be verified as follows. I tested it with FramePack video generation on both Windows and Linux.

```sh
mkdir test
cd test
python -m venv .venv
source .venv/bin/activate
pip install torch torchvision --index-url https://download.pytorch.org/whl/cu124
pip install git+https://github.com/xhiroga/musubi-tuner-xhiroga.git@feat/package
# For uv, just: uv add git+https://github.com/xhiroga/musubi-tuner-xhiroga.git@feat/package

python -c "from musubi_tuner import fpack_generate_video; fpack_generate_video.main()"  --dit ...... (options below)
```

I hope this gives you the impression that it's surprisingly straightforward...!

※1: (There are [flat layout and src layout](https://packaging.python.org/ja/latest/discussions/src-layout-vs-flat-layout/). The latter is more common, so I went with that approach)

## About the Code

I apologize for the very large diff... I think it would be easier to understand if you look at it commit by commit.

1. Move code to `src/musubi_tuner` ([8accd4e](https://github.com/kohya-ss/musubi-tuner/commit/8accd4e5a8889b5d14f93ce54b452a76b1ce3fff))
2. Fix imports ([e6d4f7e](https://github.com/kohya-ss/musubi-tuner/commit/e6d4f7e9bf1f6c9d406207522cc02e8122e13a70))
3. Add `__init__.py` to `src/musubi_tuner` ([68ffab8](https://github.com/kohya-ss/musubi-tuner/commit/68ffab884f4b9b7a5551416a88adcc696477cdc2))
4. Fix documentation ([fa9d64d](https://github.com/kohya-ss/musubi-tuner/commit/fa9d64dae7bb17dae8923efa24b3635ef93b6e7b))

Also, this time I moved executable scripts like `fpack_train_network.py` to `src/musubi_tuner` as well. This means the paths for existing commands will change slightly...!

To avoid this, there's a compromise approach of moving only common modules like `frame_pack` and `wan` to `src/musubi_tuner`. In this case, scripts wouldn't be available via import, but please let me know if you'd prefer this approach!

## Motivation

I've been using kohya-san's code since the sd-script days, but as I fork and make repeated modifications, I sometimes find that changes that might be beneficial to others become complexly intertwined with other changes, making it difficult to submit PRs...

Using it as a package would make such situations less likely to occur, so I proposed this approach.

<details>
<summary>日本語</summary>

いつもお忙しい中、目を通していただいてありがとうございます...!

musubi-tunerのパッケージ化ですが、試しにやってみたところ思ったよりも手軽そうでした。そのためご共有用のPRです！

一旦取り込む予定なし、ということでクローズいただいても全く問題ないです！お手すきに、お気軽にご覧いただければ嬉しいです。

## パッケージ化について

パッケージ化は、実はコードを所定のレイアウト※1に移動するだけで完了します。

- `setup.py` の作成 → 不要でした。`uv` の採用に伴って、`pyproject.toml` に必要事項がすでに詰まっているため
- PyPIへの登録 → 必須ではないです。GitHubから直接 `pip install` できるため。ちなみにブランチ指定等も可能です

本PRの内容は、次のように検証できます。WindowsとLinuxでFramePackの動画生成にて検証しました。

```sh
mkdir test
cd test
python -m venv .venv
source .venv/bin/activate
pip install torch torchvision --index-url https://download.pytorch.org/whl/cu124
pip install git+https://github.com/xhiroga/musubi-tuner-xhiroga.git@feat/package
# uvの場合は uv add git+https://github.com/xhiroga/musubi-tuner-xhiroga.git@feat/package のみです

python -c "from musubi_tuner import fpack_generate_video; fpack_generate_video.main()"  --dit ...... (以降オプション)
```

ここまでで、意外と手軽そう、と思っていただけたら幸いです...!

※1: （[flat layoutとsrc layout](https://packaging.python.org/ja/latest/discussions/src-layout-vs-flat-layout/)があります。後者のほうがメジャーなのでそのようにしました）

## コードについて

差分が非常に多く、申し訳ありません...コミット単位に見ていただいた方が分かりやすいと思います。

1. `src/musubi_tuner` へのコードの移動 ([8accd4e](https://github.com/kohya-ss/musubi-tuner/commit/8accd4e5a8889b5d14f93ce54b452a76b1ce3fff))
2. importの修正 ([e6d4f7e](https://github.com/kohya-ss/musubi-tuner/commit/e6d4f7e9bf1f6c9d406207522cc02e8122e13a70))
3. `src/musubi_tuner` への `__init__.py` の追加 ([68ffab8](https://github.com/kohya-ss/musubi-tuner/commit/68ffab884f4b9b7a5551416a88adcc696477cdc2))
4. ドキュメントの修正 ([fa9d64d](https://github.com/kohya-ss/musubi-tuner/commit/fa9d64dae7bb17dae8923efa24b3635ef93b6e7b))

また、今回は `fpack_train_network.py` などの実行用スクリプトも `src/musubi_tuner` に移動しています。このため、既存のコマンドのパスが少し変わってしまうことになります...!

これを避けるために、`frame_pack` や `wan` などの共通モジュールだけを `src/musubi_tuner` に移動する、という折衷案もあります。この場合スクリプトはimportで使えなくなりますが、こちらの方がよろしければ遠慮なく仰ってください！

## 動機

訓練や推論をアプリケーションに組み込む際、git cloneを含む複雑なコードになりがちなので、それを解消できると非常に助かります...!

それ以外にも、sd-scriptの話なのですが、フォークして改変を重ねているうちに、他の方にも有益そうな変更がそれ以外の変更と複雑に絡み合ってPRを出しづらいことがありまして...パッケージとしての利用ならそうしたことが起きづらいと考えています。

</details>

JCBrouwer commented:
I would really appreciate this! Makes it much easier to use and extend musubi tuner :)

xhiroga commented:
Thank you for your confirmation!

> I apologize for confirming again, but in this case, if a specific version of PyTorch has been installed before, will `pip install -e .` overwrite that version of PyTorch with the version of PyTorch described in pyproject.toml?

Yes, basically no. Since we have removed torch from the dependencies in `pyproject.toml` and moved it to `extra`, it will not be referenced by `pip install -e . . `, because it is not referenced by `pip install -e .

(As an extreme case, you can run `pip install torch torchvision --index-url https://download.pytorch.org/whl/**cu116 (or other old one)**` -> `pip install -e . `, etc., will result in the following errors: 1. torch 1.X is installed 2. `accelerate` or `bitsandbytes` depends on `torch 2.X.' This still occurs in the current situation).

Other points that may possibly confuse users are as follows... (This still occurs in the current situation.) Other points that may confuse users are as follows.
(This is still the case.) Other points that may be confusing to users are as follows.

- If you specify "extra" in pip, you will get an error, even though "extra" is required in uv (because index-url in pyproject is specified only for uv).
- In case of uv, use `uv run src/musubi_tuner/... `, but `uv run --extra cu124 src/musubi_tuner/... ` instead of `uv run src/musubi_tuner/...`. It may take some getting used to.

<details>
<summary>日本語</summary>

ご確認ありがとうございます！

> 再度の確認ですが、この場合、先にPyTorchの特定バージョンをインストールしていた時、`pip install -e .` により、そのPyTorchがpyproject.tomlに記述されているバージョンのPyTorchで上書きされてしまうことはありませんでしょうか。

はい、ありません。`pyproject.toml`のdependencyからtorchを外してextraに移動しているので、`pip install -e .`では参照されないためです。

（極端なケースとして、`pip install torch torchvision --index-url https://download.pytorch.org/whl/cu116`(など古いもの) → `pip install -e .` などを実行すると、1. torch 1.Xがインストールされる 2. `accelerate`や`bitsandbytes`が torch 2.Xに依存している、でエラーになります。しかし、これは現状でも発生します）

その他、もしかしたらユーザーが混乱するかも...という点は次のとおりです。
現状、対策はドキュメントでの案内ということになると思います。

- uvではextraの指定が必要なのに、pipでextraを指定するとエラーになります（pyproject内の index-url の指定方法が uv 専用のため）
- `uv sync` だけでなく `run` においても `uv run src/musubi_tuner/...` ではなく `uv run --extra cu124 src/musubi_tuner/...` という指定が必須です。仕様なんですが、ちょっと慣れが必要かもしれません。

</details>

---

<details>
<summary>pip install -e . 動作確認ログ</summary>

```console
 hiroga@HIROGA-RTX4090  musubi-tuner  feat/package  % rm -rf .venv

 hiroga@HIROGA-RTX4090  musubi-tuner  feat/package  % python3.10 -m venv .venv

 hiroga@HIROGA-RTX4090  musubi-tuner  feat/package  % .venv/bin/pip install torch torchvision --extra-index-url https://download.pytorch.org/whl/cu128 
Looking in indexes: https://pypi.org/simple, https://download.pytorch.org/whl/cu128
Collecting torch
  Using cached https://download.pytorch.org/whl/cu128/torch-2.7.1%2Bcu128-cp310-cp310-manylinux_2_28_x86_64.whl (1039.4 MB)
Collecting torchvision
  Using cached https://download.pytorch.org/whl/cu128/torchvision-0.22.1%2Bcu128-cp310-cp310-manylinux_2_28_x86_64.whl (8.7 MB)
...
Installing collected packages: nvidia-cusparselt-cu12, mpmath, typing-extensions, triton, sympy, pillow, nvidia-nvtx-cu12, nvidia-nvjitlink-cu12, nvidia-nccl-cu12, nvidia-curand-cu12, nvidia-cufile-cu12, nvidia-cuda-runtime-cu12, nvidia-cuda-nvrtc-cu12, nvidia-cuda-cupti-cu12, nvidia-cublas-cu12, numpy, networkx, MarkupSafe, fsspec, filelock, nvidia-cusparse-cu12, nvidia-cufft-cu12, nvidia-cudnn-cu12, jinja2, nvidia-cusolver-cu12, torch, torchvision
Successfully installed MarkupSafe-3.0.2 filelock-3.18.0 fsspec-2025.5.1 jinja2-3.1.6 mpmath-1.3.0 networkx-3.4.2 numpy-2.2.6 nvidia-cublas-cu12-12.8.3.14 nvidia-cuda-cupti-cu12-12.8.57 nvidia-cuda-nvrtc-cu12-12.8.61 nvidia-cuda-runtime-cu12-12.8.57 nvidia-cudnn-cu12-9.7.1.26 nvidia-cufft-cu12-11.3.3.41 nvidia-cufile-cu12-1.13.0.11 nvidia-curand-cu12-10.3.9.55 nvidia-cusolver-cu12-11.7.2.55 nvidia-cusparse-cu12-12.5.7.53 nvidia-cusparselt-cu12-0.6.3 nvidia-nccl-cu12-2.26.2 nvidia-nvjitlink-cu12-12.8.61 nvidia-nvtx-cu12-12.8.55 pillow-11.2.1 sympy-1.14.0 torch-2.7.1+cu128 torchvision-0.22.1+cu128 triton-3.3.1 typing-extensions-4.14.0

 hiroga@HIROGA-RTX4090  musubi-tuner  feat/package  % ls .venv/lib/python3.10/site-packages | grep torch                                              
functorch
torch
torch-2.7.1+cu128.dist-info
torchgen
torchvision
torchvision-0.22.1+cu128.dist-info
torchvision.libs

 hiroga@HIROGA-RTX4090  musubi-tuner  feat/package  % .venv/bin/pip install -e .                                                                      
Obtaining file:///home/hiroga/Documents/GitHub/musubi-tuner
...
Successfully installed accelerate-1.7.0 av-14.0.1 bitsandbytes-0.46.0 certifi-2025.4.26 charset-normalizer-3.4.2 diffusers-0.33.1 easydict-1.13 einops-0.8.1 ftfy-6.3.1 hf-xet-1.1.3 huggingface-hub-0.32.4 idna-3.10 importlib-metadata-8.7.0 musubi-tuner-0.1.0 opencv-python-4.11.0.86 packaging-25.0 psutil-7.0.0 pyyaml-6.0.2 regex-2024.11.6 requests-2.32.3 safetensors-0.5.3 sageattention-1.0.6 tokenizers-0.21.1 toml-0.10.2 tqdm-4.67.1 transformers-4.52.4 urllib3-2.4.0 voluptuous-0.15.2 wcwidth-0.2.13 zipp-3.22.0

 hiroga@HIROGA-RTX4090  musubi-tuner  feat/package  % ls .venv/lib/python3.10/site-packages | grep torch                                              
functorch
torch
torch-2.7.1+cu128.dist-info
torchgen
torchvision
torchvision-0.22.1+cu128.dist-info
torchvision.libs

```

<details>

maybleMyers commented:
I am mixed of opinions for it. It seems to make it harder to edit the code and with video models changing so frequently it might be a bad thing. If it takes time away from kohya to implement then it is a bad thing for sure. Right now the important thing is to focus on new features and supporting various new models.

I do agree that we need some type of framework to add to the original repository but it kind of exists, you would just need to add a new script ie phanton_generate_video.py for example. For me I do not have enough time and all models require so many change to support.

maybleMyers commented:
Don't worry about the impact on me :D 
I usually refactor everything to work with my changes anyways. I probably should stick to a better format though.

xhiroga commented on src/musubi_tuner/wan_generate_video.py:
```diff
@@ -21,27 +21,27 @@
+from musubi_tuner.networks import lora_wan
+from musubi_tuner.utils.safetensors_utils import mem_eff_save_file, load_safetensors
+from musubi_tuner.wan.configs import WAN_CONFIGS, SUPPORTED_SIZES
+import musubi_tuner.wan
```
誤りです。 https://github.com/kohya-ss/musubi-tuner/pull/342 で修正しました。

xhiroga commented on src/musubi_tuner/wan_cache_text_encoder_outputs.py:
```diff
@@ -6,20 +6,20 @@
-from wan.configs import wan_t2v_14B
+from musubi_tuner.wan.configs import wan_t2v_14B
 
-import cache_text_encoder_outputs
```
誤りです。 https://github.com/kohya-ss/musubi-tuner/pull/342 で修正しました。

xhiroga commented on src/musubi_tuner/wan_cache_latents.py:
```diff
@@ -7,18 +7,18 @@
+from musubi_tuner.wan.configs import wan_i2v_14B
+from musubi_tuner.wan.modules.vae import WanVAE
+from musubi_tuner.wan.modules.clip import CLIPModel
+import musubi_tuner.cache_latents
```
誤りです。 https://github.com/kohya-ss/musubi-tuner/pull/342 で修正しました。

xhiroga reviewed (commented):
In creating the wrapper script, we found an error in import and corrected it at https://github.com/kohya-ss/musubi-tuner/pull/342. We have searched for similar errors.
We apologize for the oversight!

<details>
<summary>日本語</summary>
ラッパースクリプト作成にあたってimportの誤りを見つけたため、 https://github.com/kohya-ss/musubi-tuner/pull/342 で修正しました。同様の誤りがないことは検索済みです。
見落としがあり、申し訳ありません！
</details>

xhiroga commented:
> I would like to make the following modifications to pyproject.toml after merging, but will this be okay for the packaging?

So that was it. Yes, no problem at all! Thank you very much!
If you run `uv sync --extra cu124` after the change and there are no errors, I think you are good to go.

<details>
<summary>日本語</summary>

> pyproject.tomlについて、以下のような修正をマージ後に加えたいと思いますが、パッケージとしては問題ないでしょうか。
（pyproject.tomlはconributeしていただいたもので私はノータッチだったので、ここで既存のrequirements.txtと合わせておきたいと思います。）

そうだったのですね。はい、全く問題ありません！ありがとうございます！
変更後に `uv sync --extra cu124`を実行してエラーがなければ問題ないと思います。

</details>

xhiroga commented:
ありがとうございます！
お手数をおかけいたします！
